File: sections/66_master_math_mixed_contest/problems/c_wildcard_rotation/solution.cpp
#include <bits/stdc++.h>
#include "../../../65_polynomial_algorithms_recurrences/ntt.hpp"
using namespace std;
using course_ntt::convolution;

vector<int> correlation(const vector<int> &left, const vector<int> &right) {
    int n = left.size();
    vector<int> reversed(n), doubled(2 * n);
    for (int i = 0; i < n; ++i) {
        reversed[n - 1 - i] = left[i];
        doubled[i] = doubled[i + n] = right[i];
    }
    vector<int> product = convolution(reversed, doubled);
    vector<int> result(n);
    for (int shift = 0; shift < n; ++shift)
        result[shift] = product[n - 1 + shift];
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    string s, t;
    if (!(cin >> n >> s >> t)) return 0;
    vector<int> concrete_s(n), concrete_t(n);
    for (int i = 0; i < n; ++i) {
        concrete_s[i] = s[i] != '?';
        concrete_t[i] = t[i] != '?';
    }
    vector<int> both = correlation(concrete_s, concrete_t), matches(n);
    for (char letter : string("ACGT")) {
        vector<int> in_s(n), in_t(n);
        for (int i = 0; i < n; ++i) {
            in_s[i] = s[i] == letter;
            in_t[i] = t[i] == letter;
        }
        vector<int> same = correlation(in_s, in_t);
        for (int shift = 0; shift < n; ++shift) matches[shift] += same[shift];
    }
    int count = 0, first = -1;
    for (int shift = 0; shift < n; ++shift)
        if (both[shift] == matches[shift]) {
            if (first == -1) first = shift;
            ++count;
        }
    cout << count << ' ' << first << '\n';
}

File: sections/66_master_math_mixed_contest/problems/c_wildcard_rotation/solution.py
import sys
from pathlib import Path

SECTION_65 = Path(__file__).resolve().parents[3] / "65_polynomial_algorithms_recurrences"
sys.path.insert(0, str(SECTION_65))
from ntt import convolution


def correlation(left: list[int], right: list[int]) -> list[int]:
    n = len(left)
    product = convolution(left[::-1], right + right)
    return product[n - 1 : 2 * n - 1]


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    s, t = data[1].decode(), data[2].decode()
    both = correlation([c != "?" for c in s], [c != "?" for c in t])
    matches = [0] * n
    for letter in "ACGT":
        same = correlation([c == letter for c in s], [c == letter for c in t])
        for shift in range(n):
            matches[shift] += same[shift]
    valid = [shift for shift in range(n) if both[shift] == matches[shift]]
    print(len(valid), valid[0] if valid else -1)


if __name__ == "__main__":
    main()
